File: backend_python/blockchain/blockchain.py
import time, json, os
from .block import Block

DATA_FILE = "data/blockchain.json"


class Blockchain:
    def __init__(self):
        self.chain = []
        self.voted_hashes = set()
        self.load()

    def create_genesis_block(self):
        return Block(0, "0", int(time.time()), votes=[])

    def get_latest_block(self):
        return self.chain[-1]

    def has_voted(self, voter_hash: str) -> bool:
        return voter_hash in self.voted_hashes

    def add_block(self, votes):
        """
        votes: List[{
            voter_hash: str,
            proof: str,
            proof_valid: bool
        }]
        """

        # proof_valid チェック
        for v in votes:
            if not v.get("proof_valid", False):
                raise ValueError("Invalid proof detected")

        # 1人1票チェック（voter_hash）
        for v in votes:
            if v["voter_hash"] in self.voted_hashes:
                raise ValueError("This voter has already voted")

        prev_block = self.get_latest_block()

        new_block = Block(
            index=prev_block.index + 1,
            previous_hash=prev_block.hash,
            timestamp=int(time.time()),
            votes=votes
        )

        self.chain.append(new_block)

        # 投票済みとして記録
        for v in votes:
            self.voted_hashes.add(v["voter_hash"])

        self.save()
        return new_block

    def save(self):
        data = {
            "chain": [b.to_dict() for b in self.chain],
            "voted_hashes": list(self.voted_hashes)
        }
        os.makedirs("data", exist_ok=True)
        with open(DATA_FILE, "w") as f:
            json.dump(data, f, indent=4)

    def load(self):
        if os.path.exists(DATA_FILE):
            with open(DATA_FILE, "r") as f:
                data = json.load(f)
                self.chain = [Block.from_dict(b) for b in data.get("chain", [])]
                self.voted_hashes = set(data.get("voted_hashes", []))
        else:
            genesis = self.create_genesis_block()
            self.chain = [genesis]
            self.voted_hashes = set()
            self.save()

File: backend_go/blockchain/blockchain.go
package blockchain

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"os"
	"sync"
	"time"
	"fmt"
)

type Transaction struct {
	Voter     string `json:"voter"`
	Vote      string `json:"vote"`
	Signature string `json:"signature"`
}

type Block struct {
	Index        int           `json:"index"`
	PreviousHash string        `json:"previous_hash"`
	Timestamp    int64         `json:"timestamp"`
	Votes        []Transaction `json:"votes"`
	Nonce        int           `json:"nonce"`
	Hash         string        `json:"hash"`
}

type Blockchain struct {
	Chain           []Block
	VotedPublicKeys map[string]struct{}
	mu              sync.Mutex
}

const DataFile = "data/blockchain.json"

var BlockchainInstance = NewBlockchain()

func NewBlockchain() *Blockchain {
	bc := &Blockchain{
		Chain:           []Block{},
		VotedPublicKeys: make(map[string]struct{}),
	}
	bc.Load()

	if len(bc.Chain) == 0 {
		genesis := bc.createGenesisBlock()
		bc.Chain = []Block{genesis}
		bc.Save()
	}
	return bc
}

func (bc *Blockchain) createGenesisBlock() Block {
	b := Block{
		Index:        0,
		PreviousHash: "0",
		Timestamp:    time.Now().Unix(),
		Votes:        []Transaction{},
		Nonce:        0,
	}
	b.Hash = b.calculateHash()
	return b
}

func (b *Block) calculateHash() string {
	voteBytes, _ := json.Marshal(b.Votes)
    record := fmt.Sprintf("%d%s%d%s%d", b.Index, b.PreviousHash, b.Timestamp, voteBytes, b.Nonce)
    sum := sha256.Sum256([]byte(record))
	return hex.EncodeToString(sum[:])
}

func (bc *Blockchain) GetLatestBlock() Block {
	bc.mu.Lock()
	defer bc.mu.Unlock()
	return bc.Chain[len(bc.Chain)-1]
}

func (bc *Blockchain) AddBlock(votes []Transaction) (Block, error) {
    bc.mu.Lock()
    defer bc.mu.Unlock()

    for _, v := range votes {
        if _, ok := bc.VotedPublicKeys[v.Voter]; ok {
            return Block{}, errors.New("this public key has already voted")
        }
    }

    prev := bc.Chain[len(bc.Chain)-1]
    newBlock := Block{
        Index:        prev.Index + 1,
        PreviousHash: prev.Hash,
        Timestamp:    time.Now().Unix(),
        Votes:        votes,
        Nonce:        0,
    }
    newBlock.Hash = newBlock.calculateHash()

    bc.Chain = append(bc.Chain, newBlock)

    for _, v := range votes {
        bc.VotedPublicKeys[v.Voter] = struct{}{}
    }

    go func() {
		_ = bc.Save()
	}()

    return newBlock, nil
}

func (bc *Blockchain) HasVoted(pubKey string) bool {
	bc.mu.Lock()
	defer bc.mu.Unlock()
	_, ok := bc.VotedPublicKeys[pubKey]
	return ok
}

func (bc *Blockchain) GetResults() map[string]int {
	bc.mu.Lock()
	defer bc.mu.Unlock()
	results := make(map[string]int)
	for _, block := range bc.Chain {
		for _, v := range block.Votes {
			results[v.Vote]++
		}
	}
	return results
}

func (bc *Blockchain) GetChain() []Block {
	bc.mu.Lock()
	defer bc.mu.Unlock()
	return bc.Chain
}

func (bc *Blockchain) Save() error {
    bc.mu.Lock()
    defer bc.mu.Unlock()

    os.MkdirAll("data", os.ModePerm)
    file, err := os.Create(DataFile)
    if err != nil {
        return err
    }
    defer file.Close()

    data := struct {
        Chain           []Block  `json:"chain"`
        VotedPublicKeys []string `json:"voted_public_keys"`
    }{
        Chain: bc.Chain,
    }

    for k := range bc.VotedPublicKeys {
        data.VotedPublicKeys = append(data.VotedPublicKeys, k)
    }

    enc := json.NewEncoder(file)
    enc.SetIndent("", "  ")
    err = enc.Encode(data)

    return err
}

func (bc *Blockchain) Load() error {
	bc.mu.Lock()
	defer bc.mu.Unlock()

	file, err := os.Open(DataFile)
	if err != nil {
		return nil
	}
	defer file.Close()

	data := struct {
		Chain           []Block  `json:"chain"`
		VotedPublicKeys []string `json:"voted_public_keys"`
	}{}

	dec := json.NewDecoder(file)
	if err := dec.Decode(&data); err != nil {
		return err
	}

	bc.Chain = data.Chain
	bc.VotedPublicKeys = make(map[string]struct{})
	for _, k := range data.VotedPublicKeys {
		bc.VotedPublicKeys[k] = struct{}{}
	}

	return nil
}
